pandas/_libs/tslibs/conversion.pyx

```diff
@@ -742,11 +747,18 @@ cdef datetime _localize_pydatetime(datetime dt, tzinfo tz):
         identically, i.e. discards nanos from Timestamps.
         It also assumes that the `tz` input is not None.
     """
-    try:
-        # datetime.replace with pytz may be incorrect result
-        # TODO: try to respect `fold` attribute
-        return tz.localize(dt, is_dst=None)
-    except AttributeError:
+    if treat_tz_as_pytz(tz):
+        import pytz
+        try:
+            # datetime.replace with pytz may be incorrect result
+            # TODO: try to respect `fold` attribute
+            return tz.localize(dt, is_dst=None)
+        except (pytz.AmbiguousTimeError, pytz.NonExistentTimeError) as err:
+            if isinstance(err, pytz.AmbiguousTimeError):
+                raise AmbiguousTimeError(str(err)) from err
```

Comment:
What does raise...from even do here? Can't we just not catch them to begin with and get the same behavior?

Reply:
My thinking here is that we needed to replace raising `pytz` exceptions with our own versions elsewhere (when we try to infer DST) so here I'm also re-raising a pandas exception instead of a pytz one. Do you think that's overkill here?

Reply:
Hmm yea I am a little unwary of doing this; by providing compat here we then are forcing pandas to add these exceptions to its API no? Over the long run I'm not sure that is worth it

Reply:
I think we need to make decisions about how to handle `fold` before this.  IIUC zoneinfo doesn't have its own analogues to these exceptions bc it expects fold to always be there to resolve potential ambiguity (not sure about NonExistent).  So if we are changing over to zoneinfo, then i expect that our scalar Timestamp (which has fold) to never face ambiguity, while DatetimeIndex (which doesnt have fold) will still have ambiguity.  Ideally we'd get the scalar and vector ops to behave the same, but adding fold to DTI/DTA wont be trivial.

Reply:
> by providing compat here we then are forcing pandas to add these exceptions to its API no? 

Yeah, I guess since this is for 3.0 we can just change the exception type.